Classes/Domain/Repository/ClientRepository.php

```diff
@@ -50,45 +44,62 @@ public function __construct()
      */
     public function getClientEntity($clientIdentifier, $grantType = null, $clientSecret = null, $mustValidateSecret = true)
     {
-        $clientData = $this->findRawByIdentifier($clientIdentifier, ['*']);
-        if (!$clientData) {
+        // not all steps in OAuth2 supply a client secret, so we only validate it if it is supplied
+        if ($mustValidateSecret && $clientSecret !== null && !$this->validateClient($clientIdentifier, $clientSecret, $grantType)) {
+            $this->logger->debug(sprintf('Submitted secret "%s" is invalid for client %s', $clientSecret, $clientIdentifier));
+
             return null;
         }
 
-        // not all steps in OAuth2 supply a client secret, so we only validate it if it is supplied
-        if ($mustValidateSecret && $clientSecret !== null) {
-            if (!$this->validateClient($clientData, (string)$clientSecret)) {
-                $this->logger->debug(sprintf('Submitted secret "%s" is invalid for client %s', (string)$clientSecret, $clientIdentifier));
-
-                return null;
-            }
+        $clientData = $this->findRawByIdentifier($clientIdentifier);
+        if (!$clientData) {
+            return null;
         }
 
         return new Client(
             $clientIdentifier,
             $clientData['name'],
-            GeneralUtility::trimExplode("\n", $clientData['redirect_uris'])
+            GeneralUtility::trimExplode("\n", $clientData['redirect_uris']),
+            true // todo: make configurable: $clientData['confidential']
         );
     }
 
-    protected function findRawByIdentifier(string $clientIdentifier, $selects = ['*']): ?array
+    protected function findRawByIdentifier(string $clientIdentifier): ?array
     {
         return $this->queryBuilder
             ->resetQueryParts()
-            ->select(...$selects)
-            ->from(static::TABLE_NAME)
+            ->select('*')
+            ->from(self::TABLE_NAME)
             ->where($this->queryBuilder->expr()->eq(
                 'identifier',
                 $this->queryBuilder->createNamedParameter($clientIdentifier)
             ))
             ->setMaxResults(1)
             ->execute()
-            ->fetch() ?: null;
+            ->fetchAssociative() ?: null;
     }
 
-    private function validateClient(array $clientData, string $submittedSecret)
+    /**
+     * Validate a client's secret.
+     *
+     * @param string $clientIdentifier The client's identifier
+     * @param null|string $clientSecret The client's secret (if sent)
+     * @param null|string $grantType The type of grant the client is using (if sent)
```

Comment:
If we're introducing new methods/opening them to the public, the type hints should IMO be part of the method signature.

And I also think a (public) method with this name does not belong to a repository.

Reply:
I just realized that this method comes from the interface. Gnah.